src/planet-a/SubdivTriangleMesh.h

```diff
@@ -0,0 +1,487 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "SubdivSkeleton.h"
+
+#include <osp/types.h>
+
+#include <memory>
+#include <variant>
+
+namespace planeta
+{
+
+// ID for all chunks, from 0 to m_chunkMax
+enum class ChunkId : uint16_t {};
+
+// ID for all shared vertices; from 0 to m_sharedMax
+enum class SharedVrtxId : uint32_t {};
+
+// Indices of vertices, unaware of vertex size;
+// from 0 to (m_chunkMax*m_chunkVrtxCount + m_sharedMax)
+enum class VertexId : uint32_t {};
+
+// IDs for any chunk's shared vertices; from 0 to m_chunkVrtxSharedCount
+enum class ChunkLocalSharedId : uint16_t {};
+
+// IDs for any chunk's fill vertices; from 0 to m_chunkVrtxSharedCount
+enum class ChunkLocalFillId : uint16_t {};
+
+using SharedVrtxStorage_t = osp::IdRefCount<SharedVrtxId>::Storage_t;
+
+//-----------------------------------------------------------------------------
+
+/**
+ * Convert XY coordinates to a triangular number index
+ *
+ *  0
+ *  1  2
+ *  3  4  5
+ *  6  7  8  9
+ * 10 11 12 13 14
+ * ...
+ * x = right; y = down; (0, 0) = 0
+ *
+ * @param x [in]
+ * @param y [in]
+ * @return triangular number
+ */
+constexpr int xy_to_triangular(int x, int y) noexcept
+{
+    return y * (y + 1) / 2 + x;
+};
+
+constexpr std::pair<ChunkLocalSharedId, bool> coord_to_shared(
+        uint16_t x, uint16_t y, uint16_t chunkEdgeVrtxCount) noexcept
```

Comment:
I'm not a fan of mixing 3 different fixed size integer types.

You've got uint16_t for the x and y, then internal to the function you're storing them as an intermediate value as uint32_t, then converting the uint32_t to ChunkLocalSharedId, which is uint16_t again.

maybe do

std::underlying_type_t<ChunkLocalSharedId> x, std::underlying_type_t<ChunkLocalSharedId> y, std::underlying_type_t<ChunkLocalSharedId> chunkEdgeVrtxCount

and do the same for the offset variable inside the function.

Reply:
I'm confused over what you're suggesting with the std::underlying_type_t. offset being uint32_t is leftover from when ChunkLocalSharedId was 32-bit, and can just be changed to uint16_t as well. Again, how about uint_fast16_t?